spec/services/course/user_registration_service_spec.rb

```diff
@@ -143,5 +143,18 @@ def self.registration_with_registration_code
         expect(registration.course_user).to be_approved
       end
     end
+
+    describe '#assign_email_to_user' do
+      registration_with_invitation_code
+      context 'when the email already belongs to another user' do
+        before { invitation.user_email.update(user: create(:user)) }
+
+        it 'raises an IllegalStateError' do
+          invitation.save!
+          expect { subject.send(:accept_invitation, registration, invitation) }.
+            to raise_error(IllegalStateError)
```

Comment:
It would be better to have a feature spec, where email is taken, the user should still be able to register.